### PR TITLE
fix: Unwrap subject returned by cy.then

### LINKED ISSUE #1084: Cypress type definitions don't unwrap subject in `cy.then`
The current cypress types don't properly handle returning a thenable from `cy.then`. It will return the wrapped subject instead.

- Operating System: N/A
- Cypress Version: 1.2
- Browser Version: N/A

### Is this a Feature or Bug?
Bug

### Current behavior:
```ts
cy.wrap('foo').then(subject => {
  subject // $ExpectType string
  return cy.wrap(subject)
}).then(subject => {
  subject // $ExpectType Chainable<string>
})
```

### Desired behavior:
```ts
cy.wrap('foo').then(subject => {
  subject // $ExpectType string
  return cy.wrap(subject)
}).then(subject => {
  subject // $ExpectType string
})
```

#### Test code:
In `cli/types/tests/cypress-tests.ts`
```ts
cy.wrap('foo').then(subject => {
  subject // $ExpectType string
  return cy.wrap(subject)
}).then(subject => {
  subject // $ExpectType string
})
```
In the `cli` folder run:
```
npm run dtslint
```

### Additional Info (images, stack traces, etc)

I have a fix pending, but I wanted to log it as an issue first.

### PR DESCRIPTION
**Pull request by NicholasBoll**
Fixes #1084

### CODE REVIEW

**Comment by bahmutov** [comment]
seems correct, thanks for this fix!

**Comment by NicholasBoll** [comment]
I see that a few CI builds failed. Looking into them doesn't seem related to my changes. The type definition changes should effect any runtime tests. Is there anything I need to do on my end?

File: cli/types/index.d.ts [hunk 1]
```diff
@@ -720,6 +720,8 @@ declare namespace Cypress {
      *
      * @see https://on.cypress.io/then
      */
+    then<S extends object | any[] | string | number | boolean>(fn: (this: ObjectLike, currentSubject: Subject) => Chainable<S>, options?: Partial<Timeoutable>): Chainable<S>
+    then<S extends object | any[] | string | number | boolean>(fn: (this: ObjectLike, currentSubject: Subject) => PromiseLike<S>, options?: Partial<Timeoutable>): Chainable<S>
```
Comment (NicholasBoll) [inline cli/types/index.d.ts]: An option was coming up with a custom `Thenable` interface to match both uses cases, but I figured I'd just use standard ones (we define `Chainable` and TypeScript defines `PromiseLike` for all Spec compliant Promise implementations (including Bluebird)).

File: cli/types/tests/cypress-tests.ts [hunk 2]
```diff
@@ -138,3 +138,17 @@ cy.wrap([{ foo: 'bar' }, { foo: 'baz' }])
...
+cy.wrap('foo').then(subject => {
+  subject // $ExpectType string
+  return cy.wrap(subject)
+}).then(subject => {
+  subject // $ExpectType string
+})
+
+cy.wrap('foo').then(subject => {
+  subject // $ExpectType string
+  return Promise.resolve(subject)
```
Comment (NicholasBoll) [inline cli/types/tests/cypress-tests.ts]: Maybe I should use `Cypress.Promise` instead?

**Comment by brian-mann** [comment]
No. Whenever a PR is submitted I generally look at the tests to ensure nothing is related to them. For some reason we have been seeing higher failure rates lately but haven't had time to look at them. There are some known flaky tests and then there are tests that have begun to be flaky - but the cause is unknown.

I usually just retry the specific failed builds to weed these out. So no - nothing needs to be done on your end.

File: cli/types/tests/cypress-tests.ts (hunk 2 above)
Comment (brian-mann) [reply cli/types/tests/cypress-tests.ts]: That would be better, yes.

**Comment by NicholasBoll** [comment]
lol. I was just about to commit the change to use `Cypress.Promise`. You're always 1 step ahead of me @brian-mann. I was also looking at different promise types like `Q.defer` and `$.Deferred`, but I would have to recommend not using those in TypeScript since they are much harder to type. Bluebird is included and spec compliant and automatically typed through inference.

**Pull request merged**